### PR TITLE
Add tests (and a bit of cleanup) for interior mut handling in promotion and const-checking

### PR DESCRIPTION
**Pull request by RalfJung**
Basically these are the parts of https://github.com/rust-lang/rust/pull/121786 that can be salvaged.

r? @oli-obk

### CODE REVIEW

File: tests/ui/consts/enclosing-scope-rule.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// as the surrounding block (which in this case is the entire program). There are multiple
+// allocations being interned at once.
+
+struct Gen<T>(T);
+impl<'a, T> Gen<&'a T> {
+    // Can't be promoted because `T` might not be `'static`.
+    const C: &'a [T] = &[];
+}
+
+// Can't be promoted because of `Drop`.
```
Comment (oli-obk) [inline tests/ui/consts/enclosing-scope-rule.rs:13]: These comments are wrong

**Comment by oli-obk** [comment]
r=me with comment nit

File: tests/ui/consts/enclosing-scope-rule.rs (hunk 1 above)
Comment (RalfJung) [reply tests/ui/consts/enclosing-scope-rule.rs:13]: Why? I think it is correct?

Comment (RalfJung) [reply tests/ui/consts/enclosing-scope-rule.rs:13]: @oli-obk what do you think is wrong about these comments? They look right to me.

Comment (oli-obk) [reply tests/ui/consts/enclosing-scope-rule.rs:13]: Hmm... I think I thought these were accidentally copy pasted from the other PR. They are correct after all

File: tests/ui/consts/refs-to-cell-in-final.rs [hunk 2]
```diff
@@ -15,4 +15,22 @@ static RAW_SYNC_S: SyncPtr<Cell<i32>> = SyncPtr { x: &Cell::new(42) };
...
+    Object(Cell<bool>),
+}
+impl Drop for JsValue {
+    fn drop(&mut self) {}
+}
+const UNDEFINED: &JsValue = &JsValue::Undefined;
+//~^ERROR: mutable pointer in final value of constant
+
+// In contrast, this one works since it is being promoted.
+const NONE: &'static Option<Cell<i32>> = &None;
```
Comment (oli-obk) [inline tests/ui/consts/refs-to-cell-in-final.rs:34]: Maybe show off the promotion somewhere other than the trailing expression?

**Review by oli-obk — APPROVED** [review]

File: tests/ui/consts/refs-to-cell-in-final.rs (hunk 2 above)
Comment (RalfJung) [reply tests/ui/consts/refs-to-cell-in-final.rs:34]: Good point, I didn't even realize that's possible -- I will add
```rust
// Making it clear that this is promotion, not "outer scope".
const NONE_EXPLICIT_PROMOTED: &'static Option<Cell<i32>> = {
    let x = &None;
    x
};
```

Comment (RalfJung) [reply tests/ui/consts/refs-to-cell-in-final.rs:34]: Interestingly, the behavior of [this example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a9a388241c867eabf6044e85d61bbd80) changed between stable and nightly. On nightly it just says "can't drop `String` in const", on stable it additionally says "temporary value dropped while borrowed". I hope that's just suppressing two errors from the same location and not secretly actually doing promotion...

**Pull request merged**